src/htbl.ml

```diff
@@ -0,0 +1,354 @@
+(*
+ * Copyright (c) 2023, Carine Morel <carine.morel.pro@gmail.com>
+ *)
+
+module Type = struct
+  type 'b kind = Dummy | Regular of 'b
+  type key = int
+end
+
+module Llist = struct
+  include Type
+
+  type 'a marked = Last | LRemove | Normal of 'a node | Remove of 'a node
+  and 'a node = { key : key; value : 'a kind; next : 'a marked Atomic.t }
+
+  type 'a t = 'a marked Atomic.t
+
+  type 'a local = {
+    prev : 'a marked Atomic.t;
+    curr : 'a marked;
+    next : 'a marked;
+  }
+
+  let init () : 'a t = Atomic.make Last
+
+  let convert_to_normal = function
+    | (Last | Normal _) as node -> node
+    | LRemove -> Last
+    | Remove node -> Normal node
+
+  let mark_to_be_removed = function
+    | Last -> LRemove
+    | Normal node -> Remove node
+    | _ as x -> x
+
+  let rec find_loop key t prev curr =
+    match curr with
+    | Last | LRemove -> (false, { prev; curr; next = Last })
```

Comment:
Good question.

First important note : the `remove` information is for the node that precedes and not the node actually wearing it.

Then it is a question of use. For the linked list, we should never have a `LRemove` or `Remove node` in this pattern matching because the only case that could happen is at the first iteration. In any other cases, we would see it when inspecting the previous node, remove the previous node and convert the `Remove` node to a `Normal` one.  But there is no case where the first call of `find` on a a type `Llist.t` could encounter a `remove` node. So for the linked list that is an impossible case. 

For the hashtable however, the buckets point to dummy nodes in the middle of a linked list. So a dummy node can actually be marked if the node right before needs to be removed. As the `find` function begins its search at the marked dummy node, it has never encountered the node that actually needs to be removed and so can neither remove the previous node or unmarked the current one.